Beats frame refactor

Holzhaus opened the pull request:
Based on #4048.

This changes the Beats object to return the FramePos type from all methods that return positions.
I tried to keep the changes in the calling code as small as possible, which admittedly makes this code quite ugly with all this back/forth conversion. But otherwise it would be much more work and the diff would be huge.

This PR makes it possible to migrate the other features to the framepos type independently later on, which is taken care of in #4061 for example.

Holzhaus committed:
Beats: Use FramePos consistently in Beats classes

Holzhaus committed:
BeatMap: Remove unused function

Holzhaus committed:
Beats: Rename findNBeatsFromSample to findNBeatsFromPosition

The method returns a mixxx::audio::FramePos value, so mentioning the
unit in the method name is unnecessary.

Holzhaus commented:
Ready to review. Sorry the diff is so large :-/

uklotzde commented on src/analyzer/analyzerbeats.cpp:
```diff
@@ -158,7 +158,9 @@ bool AnalyzerBeats::shouldAnalyze(TrackPointer pTrack) const {
 
-    if (subVersion.isEmpty() && pBeats->findNextBeat(0) <= 0.0 &&
+    if (subVersion.isEmpty() &&
+            pBeats->findNextBeat(mixxx::audio::kStartFramePos) <=
```
How about introducing a utility function to reduce the cognitive load?

```
Beats::findFirstBeat() = Beats::findNextBeat(mixxx::audio::kStartFramePos)
```

Holzhaus replied:
469c9908f3c3de4b307a332c13de9705c3db9579

Holzhaus committed:
Beats: Add `firstBeat()` utility function

This was suggested here: https://github.com/mixxxdj/mixxx/pull/4049#discussion_r664823037

Holzhaus committed:
BpmControl: Use mixxx::kEngineChannelCount instead of hardcoding 2

uklotzde commented on src/engine/controls/cuecontrol.cpp:
```diff
@@ -1979,10 +1989,12 @@ double CueControl::quantizeCuePoint(double cuePos) {
     // The closest beat can be an unreachable  interpolated beat past the end of
     // the track.
-    if (closestBeat != -1.0 && closestBeat <= total) {
+    const double closestBeat = closestBeatPosition.toEngineSamplePos();
```
Is it safe to invoke `toEngineSamplePos()` on `kInvalidValue`? Do we have a test for this use case that verifies the expected outcome?

I am aware that the value is not used if invalid. But since we invoke this function beforehand we should verify this.

Holzhaus replied:
Not yet, this is taken care of in #4061. Also see the comment below.

uklotzde commented on src/engine/controls/loopingcontrol.cpp:
```diff
@@ -1192,15 +1211,26 @@ void LoopingControl::slotBeatLoop(double beats, bool keepStartPoint, bool enable
+        const double prevBeat = prevBeatPosition.isValid()
+                ? prevBeatPosition.toEngineSamplePos()
+                : -1;
+        const double nextBeat = nextBeatPosition.isValid()
```
If `-1`is the corresponding engine sample value for invalid frame positions we might consider to return it from `toEngineSamplePos()`. I suppose this function will disappear eventually?

This conversion appears multiple times (see also: quantizecontrol.cpp), so we should at least wrap it into a utility function instead of repeating it over and over again.

Holzhaus replied:
I agree. Unfortunately I only had the idea later.

Check out https://github.com/mixxxdj/mixxx/pull/4061/commits/6d13725c
314b1aca599409d770eb52d774fa4c69 and https://github.com/mixxxdj/mixxx/pull/4061/commits/b0be360f953189e7857ac4f599b3ca953409ba58. Both are part of https://github.com/mixxxdj/mixxx/pull/4061 which is a follow up to this PR. I'm not particularity keen on cherry-picking these commits here and causing merge conflicts. But I can do it if you want.

uklotzde commented on src/track/beatgrid.cpp:
```diff
@@ -23,72 +23,76 @@ namespace mixxx {
-        m_dCurrentSample += m_dBeatLength;
-        return beat;
+    audio::FramePos next() override {
+        const audio::FramePos beatPosition = m_currentPosition;
```
Use the opportunity to simplify this code: return m_currentPosition + m_beatLengthFrames

Holzhaus replied:
no, then the iterator would not advance and always return the same position.

uklotzde replied:
Sorry, confused the variables.

uklotzde commented on src/track/beats.cpp:
```diff
@@ -2,54 +2,55 @@
+int Beats::numBeatsInRange(audio::FramePos startPosition, audio::FramePos endPosition) const {
+    audio::FramePos lastPosition = audio::kStartFramePos;
+    int i = 1;
+    for (; lastPosition < endPosition; i++) {
```
Let' use a while loop

Holzhaus replied:
56e2683741651595ef356f2cf1fa3080d0515a68

uklotzde reviewed (changes requested):
Some thoughts and suggestions.

Holzhaus committed:
Beats: Use while loop in numBeatsInRange()

uklotzde commented:
If someone wants to double check then please do. Otherwise let's merge it soon.

uklotzde commented:
If anyone finds an issue afterwards it should be trivial to fix. LGTM
